NotificationHubs/src/com/windowsazure/messaging/Installation.java

```diff
@@ -1,133 +1,366 @@
+     *
+     * @param userId The user ID for the installation.
+     */
+    public void setUserId(String userId) {
+        this.userId = userId;
+    }
+
+    /**
+     * Gets the tags for the installation.
+     *
+     * @return The tags for the installation.
+     */
+    public List<String> getTags() {
+        return tags;
+    }
+
+    /**
+     * Adds a tag to the installation.
+     *
+     * @param tag The tag to add to the installation;
+     */
+    public void addTag(String tag) {
+        if (this.tags == null) {
+            this.tags = new ArrayList<>();
+        }
+
+        this.tags.add(tag);
+    }
+
+    /**
+     * Removes a tag from the installation.
+     *
+     * @param tag The tag to remove from the installation.
+     */
+    public void removeTag(String tag) {
+        if (this.tags == null) {
+            return;
+        }
+
+        this.tags.remove(tag);
+    }
+
+    /**
+     * Clears the tags for the installation.
+     */
+    public void clearTags() {
+        if (this.tags == null) {
+            return;
+        }
+
+        this.tags.clear();
+    }
+
+    /**
+     * Gets the installation templates for the installation.
+     *
+     * @return The installation templates for the installation.
+     */
+    public Map<String, InstallationTemplate> getTemplates() {
+        return templates;
+    }
+
+    /**
+     * Adds an installation template by name to the installation.
+     *
+     * @param templateName The name for the installation template.
+     * @param template     The template to add to the installation.
+     */
+    public void addTemplate(String templateName, InstallationTemplate template) {
+        if (this.templates == null) {
+            this.templates = new HashMap<>();
+        }
+
+        this.templates.put(templateName, template);
+    }
+
+    /**
+     * Removes an installation template based upon the template name.
+     *
+     * @param templateName The name of the installation template to remove.
+     */
+    public void removeTemplate(String templateName) {
+        if (this.templates == null) {
+            return;
+        }
+
+        this.templates.remove(templateName);
+    }
+
+    /**
+     * Clears the installation templates.
+     */
+    public void clearTemplates() {
+        if (this.templates == null) {
+            return;
+        }
+
+        this.templates.clear();
+    }
+
+    /**
+     * Gets the secondary tiles for WNS
+     *
+     * @return The secondary tiles for WNS.
+     */
+    public Map<String, WnsSecondaryTile> getSecondaryTiles() {
+        return secondaryTiles;
+    }
+
+    /**
+     * Adds a secondary tile to the installation template.
+     *
+     * @param tileName The name for the tile.
+     * @param tile     THe WNS secondary tile.
+     */
+    public void addSecondaryTile(String tileName, WnsSecondaryTile tile) {
+        if (this.templates == null) {
+            this.secondaryTiles = new HashMap<>();
```

Comment:
What's the minimum version of Java that we allow? I know implicit type parameters aren't supported in older versions, but I can't remember the exact cutoff.

Reply:
I've updated this to be Java 8+